redash/query_runner/big_query.py

```diff
@@ -246,22 +250,106 @@ def _get_query_result(self, jobs, query):
+            metadata.extend(metadatum)
+            columns.extend(column)
 
         project_id = self._get_project_id()
         table_name = table_data['id'].replace("%s:" % project_id, "")
-        return {'name': table_name, 'columns': columns}
+        return {'name': table_name, 'columns': columns, 'metadata': metadata}
 
     def _get_columns_schema_column(self, column):
         columns = []
+        metadata = []
+
         if column['type'] == 'RECORD':
             for field in column['fields']:
-                columns.append(u"{}.{}".format(column['name'], field['name']))
+                col_name = u"{}.{}".format(column['name'], field['name'])
+                columns.append(col_name)
+                metadata.append({'name': col_name, 'type': field['type']})
         else:
             columns.append(column['name'])
+            metadata.append({'name': column['name'], 'type': column['type']})
+
+        return columns, metadata
+
+    def _columns_and_samples_to_dict(self, schema, samples):
+        samples_dict = {}
+        if not samples:
+            return samples_dict
+
+        # If a sample exists, its shape/length should be analogous to
+        # the schema provided (i.e their lengths should match up)
+        for i, column in enumerate(schema):
+            if column['type'] == 'RECORD':
+                associated_sample = samples[i]
+                if column.get('mode', None) == 'REPEATED':
+                    # Repeated fields have multiple samples of the same format.
+                    # We only need to show the first one as an example.
+                    associated_sample = [] if len(samples[i]) == 0 else samples[i][0]
+
+                for j, field in enumerate(column['fields']):
+                    col_name = u"{}.{}".format(column['name'], field['name'])
+                    samples_dict[col_name] = None
+                    if type(associated_sample) == list and len(associated_sample) > 0:
+                        samples_dict[col_name] = associated_sample[j]
+            else:
+                samples_dict[column['name']] = samples[i]
+
+        return samples_dict
+
+
+    def _flatten_samples(self, samples):
+        samples_list = []
+        for field in samples:
+            value = field['v']
+            if type(value) == dict:
+                samples_list.append(
+                    self._flatten_samples(value.get('f', []))
+                )
+            elif type(value) == list:
+                samples_list.append(
+                    self._flatten_samples(value)
+                )
+            else:
+                samples_list.append(value)
+
+        return samples_list
+
+    def get_table_sample(self, table_name):
+        if not self.configuration.get('loadSchema', False):
+            return {}
+
+        service = self._get_bigquery_service()
+        project_id = self._get_project_id()
+
+        dataset_id, table_id = table_name.split('.')
```

Comment:
Is there a chance that the provided table name contains more than one dot? Providing a limiting second paramter `table_name.split('.', 1)` could prevent accidential `ValueError`s during unpacking.